UefiTestingPkg/Library/FlatPageTableLib/AArch64/FlatPageTableAArch64.c

```diff
@@ -0,0 +1,371 @@
+          // 0b10 -> Access from EL0 is read-only, no write permissions at any EL
+        } else if ((ParentHeritableAttributes.Bits.ApTable & TT_TABLE_AP_MASK) == TT_TABLE_AP_NO_WRITE_ACCESS) {
+          // BIT7 toggles write access for all ELs
+          ScratchEntry.Tteb.Bits.AccessPermissions |= BIT7; // Set BIT7
+          // 0b11 -> Access from EL0 is not allowed, no write permissions at any EL
+        } else if ((ParentHeritableAttributes.Bits.ApTable & TT_TABLE_AP_MASK) == TT_TABLE_AP_MASK) {
+          ScratchEntry.Tteb.Bits.AccessPermissions |= BIT7;            // Set BIT7
+          ScratchEntry.Tteb.Bits.AccessPermissions &= ~((UINT64)BIT6); // Clear BIT6
+        }
+      }
+
+      if ((*LastEntry != NULL) &&
+          ((*LastEntry)->LinearAddress + (*LastEntry)->Length == RegionStart) &&
+          (ARM_TT_BASE_ADDRESS ((*LastEntry)->PageEntry) + (*LastEntry)->Length
+           == ARM_TT_BASE_ADDRESS (ScratchEntry.Uint64)) &&
+          (ARM_TT_BLOCK_ATTRIBUTES ((*LastEntry)->PageEntry) == ARM_TT_BLOCK_ATTRIBUTES (ScratchEntry.Uint64))
+          )
+      {
+        // Extend LastEntry.
+        (*LastEntry)->Length += RegionLength;
+      } else {
+        if (*MapCount < MapCapacity) {
+          // LastEntry points to next map entry in the array.
+          *LastEntry = &Map[*MapCount];
+        } else {
+          // LastEntry points to library internal map entry.
+          *LastEntry = OneEntry;
+        }
+
+        // Set LastEntry.
+        (*LastEntry)->LinearAddress = RegionStart;
+        (*LastEntry)->Length        = RegionLength;
+        (*LastEntry)->PageEntry     = ScratchEntry.Uint64;
+        (*MapCount)++;
+      }
+    } else {
+      ScratchEntry.Uint64 = PagingEntry[Index].Ttet.Uint64 & TT_HERITABLE_ATTRIBUTES_MASK;
+      // If the entry is a table and not the root, then pass the heritable access attributes
+      // from the parent.
+      if (Level > 0) {
+        ScratchEntry.Uint64 |= (ParentHeritableAttributes.Uint64 & TT_HERITABLE_ATTRIBUTES_MASK);
+      }
+
+      TranslationTableParseRecursive (
+        ARM_TT_BASE_ADDRESS (PagingEntry[Index].Ttet.Uint64),
+        Level + 1,
+        RegionStart,
+        ScratchEntry.Tteh,
+        Map,
+        MapCount,
+        MapCapacity,
+        LastEntry,
+        OneEntry
+        );
+    }
+  }
+}
+
+/**
+  Populate the input page/translation table map.
+
+  @param[in, out]      Map         Pointer to the PAGE_MAP struct to be populated.
+
+  @retval RETURN_SUCCESS           The translation table is parsed successfully.
+  @retval RETURN_INVALID_PARAMETER MapCount is NULL, or Map is NULL but *MapCount is nonzero.
+  @retval RETURN_BUFFER_TOO_SMALL  *MapCount is too small.
+                                   MapCount is updated to indicate the expected number of entries.
+                                   Caller may still get RETURN_BUFFER_TOO_SMALL with the new MapCount.
+**/
+EFI_STATUS

```

Comment:
nit: they are #defined to the same values, but the API claims to return an EFI_STATUS but returns a RETURN_STATUS. May as well be consistent.

Reply:
Fixed